### PR TITLE
Upstream zone-aware load balancing: support mismatched localities

### PR DESCRIPTION
**Pull request by AdamEAnderson**
Commit Message: Upstream zone-aware load balancing: support mismatched localities
Additional Description: Upstream zone-aware balancing would behave incorrectly if the upstream and local hosts sets had different localities. This PR adds support for the upstream and local host sets having both mismatched localities and different numbers of localities.
Risk Level: Very high
Testing: New tests included in PR
Docs Changes: Minimal, included
Release Notes: Included
Platform Specific Features: N/A
Runtime guard: `envoy.reloadable_features.enable_zone_routing_different_zone_counts` for the behavior change of zone-aware routing being enabled even when zone count is different. `envoy.reloadable_features.locality_routing_use_new_routing_logic` for safety reverts the entire change.
Fixes: #28419

### CODE REVIEW

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/load_balancer_impl.cc:552]: The result is the same as before (populating `state.residual_capacity_`), but we have to account for the slightly different format of the new percentages structure. The percentages includes a mapping that describes which combined locality percentages correspond to a given upstream locality index.

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/load_balancer_impl.cc:585]: `earlyExitNonLocalityRouting` needs a few minor changes:
- No longer need to disable zone-aware routing if the zone number differs, because this is now supported
- We now support the case where the upstream host set doesn't have any hosts in the local zone

I've also added an "easy out" for when zone-aware routing has no effect:
- if there are 1 or fewer local zones, then zone-aware routing won't ever change our routing decisions, so don't bother

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/load_balancer_impl.cc:637]: This method is now more complicated to determine how to interleave the localities between the local and upstream host sets even with mismatches and differing numbers of zones.

We use the `locality` field of each individual host in the hosts per locality to determine a valid interleaving, assuming that localities are already sorted properly (as described in `HostsPerLocalityImpl`).

The strategy is akin to the "merge" step in mergesort where we iteratively process entries out of each host set until all localities have been merged.

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/load_balancer_impl.cc:780]: minor change to support cross-zone routing mode with 0 local percent to route

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [inline source/common/upstream/load_balancer_impl.cc:806]: fix off-by-one bug that doesn't matter much because our precision loss from using integers is higher than this error would be.

**Comment by AdamEAnderson** [comment]
cc @wbpcode

**Comment by RyanTheOptimist** [comment]
@wbpcode friendly maintainer-on-call ping

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [inline source/common/upstream/load_balancer_impl.cc:745]: Basically I know what you want. But this code is pretty hard to follow or maintian. Could we make it more simple and clear?

Because all traffic will be routed to the upstream. So we only need to calculate the locality percentage of upstream cluster and related local locality percentage. Finally, we can output a structure like this:

```
percentages:
- upstream_percentage: xxxx
  local_percentage: xxxx
- upstream_percentage: xxxx
  local_percentage: xxxx

```

**Review by wbpcode — CHANGES_REQUESTED** [review]
Sorry for the delay. Thanks for your contribution. Only one comment to the code. Although I got your ideal and code, it's too hard to follow or maintain for future maintainer/developer.

**Comment by AdamEAnderson** [comment]
I was out last week, planning on working on it today/tomorrow. Will let you know when it's ready

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc]: used your suggestions to make it much simpler using a map

**Comment by AdamEAnderson** [comment]
@wbpcode PR should be ready for another look

**Review by wbpcode — COMMENTED** [review]
LGTM overall. Thanks so much for all your hard work and contribution.

File: source/common/upstream/load_balancer_impl.cc
Comment (wbpcode) [reply source/common/upstream/load_balancer_impl.cc:516]: hah, ignore it. I got it.

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:700]: If state.local_percent_to_route_ > 0, then the upstream host set must have at least one host in the local locality.

Basically this is making sure we aren't accidentally trying to route locally when the host set doesn't have any hosts there, as that would have us routing unintentionally to locality 0 rather than the actual local locality.

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:707]: yeah, not sure why I made this change. happy to revert

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:516]: yeah, locality 0 in the upstream host set isn't necessarily the local locality from the local host set.

**Comment by AdamEAnderson** [comment]
Made requested changes. One more thing to go over before we merge:

Right now in main there is special behavior if the number of local and upstream zones differ:
- zone-aware routing is disabled
-  the `lb_zone_number_differs` stat is incremented

After this change, neither the stat still exists, but is unused (and not incremented), and zone-aware routing is enabled.

I'm worried about:
- Users who expect zone-aware routing to be disabled in this case who won't anticipate the change in behavior
- Users who use the stat (maybe even for unrelated cases) and now this stat is dangling
- Tech debt in keeping around a stat that isn't being written to or used at all

A couple options for the stat:
1. Still increment the stat if zone number differs in `earlyExitLocalityRouting()` to keep some compatibility
2. Break cleanly on the stat and include more info in the changelog
3. Deprecate the stat and keep handling it for the time being (is there prior art for this?)
Personally I'd prefer either 2 or 3.

A couple options for the behavior change:
1. Document the change in the changelog and make no other assurances (current case)
2. Guard the behavior change behind a runtime or feature flag
Looking for input on what the policy is on changes like this.

**Comment by wbpcode** [comment]
I think a runtime flag should be good enough. Thanks for your detailed and responsible comments. 🌹

**Comment by AdamEAnderson** [comment]
Added runtime guard and updated docs/changelog to indicate the change.

**Comment by AdamEAnderson** [comment]
also added a test for the runtime guard

**Review by wbpcode — DISMISSED** [review]
LGTM. Thanks.

@mattklein123 could you take a second pass or merge when you have free time? Thanks.

File: changelogs/current.yaml
Comment (mattklein123) [inline changelogs/current.yaml:38]: I would call this a major behavior change. Can you move this to the new features section just to make it more clear that this is changing?

File: changelogs/current.yaml
Comment (mattklein123) [inline changelogs/current.yaml:82]: nit: I would probably move this into the other item, just to make it clear in one place what is changing.

File: source/common/upstream/load_balancer_impl.cc
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc:482]: This change is incredibly risky. What concerns me here is that a lot of code is changing independent of whether the feature flag is set or not. Even if it means code duplication, is there any way to split the changes out so that when the feature flag is not set there are no changes? This would make me feel more comfortable that the change can be rolled back if there is an issue. If there are no issues we can clean up the duplication later.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this long standing issue. A high level comment.

/wait

File: source/common/upstream/load_balancer_impl.cc
Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:482]: Do you have an example of another PR/change that does this? I'm mostly looking for design patterns/best practices/etc

I agree this is a good idea

File: changelogs/current.yaml
Comment (AdamEAnderson) [reply changelogs/current.yaml]: I've refactored and combined the changelog entries and moved it into the major behavior change section

lmk how this looks or if I can improve this further

File: source/common/upstream/load_balancer_impl.cc
Comment (mattklein123) [reply source/common/upstream/load_balancer_impl.cc:482]: I can't recall one recently, but I know we have done things like this in the past for super risky changes. cc @alyssawilk for thoughts.

I think the regression risk of this change is very high (so I would change that in the PR description.). This is code that has to my knowledge not changed very much since almost the very beginning of the project, and will likely effect many users. To be clear, I'm very happy you are fixing this, I just want to be careful. :)

/wait

Comment (alyssawilk) [reply source/common/upstream/load_balancer_impl.cc:482]: For high risk changes like this where you have lots of invasive bits, sometimes folks temporarily clone the entire class, but that'd suck given the inheritance model.
My best thought is clone functions which are invasively changed, latch runtime status in the base class and do things like
if (legacy_mode) {
calculateLocalityPercentagesLegacy() <- old function here
} else {
calculateLocalityPercentages() <- your new function
}

actually I think if you clone the calculateLocality and regenerateStructures functions
you'd only need if(legacy_mode)  in a few places.

Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:482]: that makes sense, thanks for the info

most of this stuff hasn't been touched in ~5 years at least, so it is definitely higher risk

Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:482]: I'll work on cloning stuff

Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc:492]: I've added a new runtime guard `envoy.reloadable_features.locality_routing_use_new_routing_logic`. Not super happy with the name, lmk if you have better suggestions.

I've split all the changed methods into legacy/current clones, where the legacy versions are completely unchanged. We latch the runtime state on construction, and use the legacy version if the runtime flag is disabled.

The only method I didn't guard is `ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts`, because it's completely separate from the rest of the stuff I've guarded and the changes are extremely minor and entirely compatible with both versions. If you'd like me to include this in the guard, it should be pretty simple.

I've changed the `load_balancer_impl_tests` to include a parameter for the legacy vs current flag so that all the tests run using both code paths. For the few tests where behavior changes depending on which path is used, I test behavior in both cases depending on what the parameter value is.

File: source/common/upstream/load_balancer_impl.cc
Comment (mattklein123) [inline source/common/upstream/load_balancer_impl.cc:469]: Super nit (sorry): Can you leave the existing code alone and then add the new code as `regenerateLocalityRoutingStructuresNew` or whatever? I think this will make the diff show no changes for the old path and only new changes. Then we can swap the names when we clean up the runtime flag later? Thank you.

/wait

Comment (AdamEAnderson) [reply source/common/upstream/load_balancer_impl.cc]: done

**Review by mattklein123 — APPROVED** [review]
Awesome, thank you!

**Pull request merged**
